src/k8s/pkg/k8sd/controllers/feature.go

```diff
@@ -69,65 +69,97 @@ func NewFeatureController(opts FeatureControllerOpts) *FeatureController {
+	go c.reconcileLoop(ctx, getClusterConfig, setFeatureStatus, "local-storage", c.triggerLocalStorageCh, c.reconciledLocalStorageCh, func(cfg types.ClusterConfig) (types.FeatureStatus, error) {
 		return features.Implementation.ApplyLocalStorage(ctx, c.snap, cfg.LocalStorage, cfg.Annotations)
 	})
 
-	go c.reconcileLoop(ctx, getClusterConfig, "metrics server", c.triggerMetricsServerCh, c.reconciledMetricsServerCh, func(cfg types.ClusterConfig) error {
+	go c.reconcileLoop(ctx, getClusterConfig, setFeatureStatus, "metrics-server", c.triggerMetricsServerCh, c.reconciledMetricsServerCh, func(cfg types.ClusterConfig) (types.FeatureStatus, error) {
 		return features.Implementation.ApplyMetricsServer(ctx, c.snap, cfg.MetricsServer, cfg.Annotations)
 	})
 
-	go c.reconcileLoop(ctx, getClusterConfig, "DNS", c.triggerDNSCh, c.reconciledDNSCh, func(cfg types.ClusterConfig) error {
-		if dnsIP, err := features.Implementation.ApplyDNS(ctx, c.snap, cfg.DNS, cfg.Kubelet, cfg.Annotations); err != nil {
-			return fmt.Errorf("failed to apply DNS configuration: %w", err)
+	go c.reconcileLoop(ctx, getClusterConfig, setFeatureStatus, "dns", c.triggerDNSCh, c.reconciledDNSCh, func(cfg types.ClusterConfig) (types.FeatureStatus, error) {
+		featureStatus, dnsIP, err := features.Implementation.ApplyDNS(ctx, c.snap, cfg.DNS, cfg.Kubelet, cfg.Annotations)
+
+		if err != nil {
+			return featureStatus, fmt.Errorf("failed to apply DNS configuration: %w", err)
 		} else if dnsIP != "" {
 			if err := notifyDNSChangedIP(ctx, dnsIP); err != nil {
-				return fmt.Errorf("failed to update DNS IP address to %s: %w", dnsIP, err)
+				// we already have featureStatus.Message which contains wrapped error of the Apply<Feature>
+				// (or empty if no error occurs). we further wrap the error to add the DNS IP change error to the message
+				changeErr := fmt.Errorf("failed to update DNS IP address to %s: %w", dnsIP, err)
+				featureStatus.Message = fmt.Sprintf("%s: %v", featureStatus.Message, changeErr)
+				return featureStatus, changeErr
 			}
 		}
-		return nil
+		return featureStatus, nil
 	})
 }
 
-func (c *FeatureController) reconcile(ctx context.Context, getClusterConfig func(context.Context) (types.ClusterConfig, error), apply func(cfg types.ClusterConfig) error) error {
+func (c *FeatureController) reconcile(
+	ctx context.Context,
+	getClusterConfig func(context.Context) (types.ClusterConfig, error),
+	apply func(cfg types.ClusterConfig) (types.FeatureStatus, error),
+	updateFeatureStatus func(context.Context, types.FeatureStatus) error,
+) error {
 	cfg, err := getClusterConfig(ctx)
 	if err != nil {
 		return fmt.Errorf("failed to retrieve cluster configuration: %w", err)
 	}
 
-	if err := apply(cfg); err != nil {
+	featureStatus, err := apply(cfg)
+	if err != nil {
 		return fmt.Errorf("failed to apply configuration: %w", err)
 	}
+
+	if err := updateFeatureStatus(ctx, featureStatus); err != nil {
+		return fmt.Errorf("failed to update feature status: %w", err)
+	}
```

Comment:
another way to write this (to avoid err being referenced by accident further down)

```go
   if status, err := apply(cfg); err != nil {
   ...
   } else if err := updateFeatureStatus(ctx, status); err != nil {
   ...
   }
```